src/Stratis.Bitcoin/Utilities/MemoryCache.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    public class MemoryCache<TKey, TValue>
+    {
+        /// <summary>Dictionary that contains cached items.</summary>
+        private readonly ConcurrentDictionary<TKey, TValue> cache;
+
+        /// <summary>Keys sorted by time they when they were accessed last time with most recent ones in the end.</summary>
+        private readonly LinkedList<TKey> keys;
+
+        /// <summary>Maximum items count that can be stored in the cache.</summary>
+        private readonly int sizeLimit;
+
+        /// <summary>Amount to compact the cache by when the maximum size is exceeded.</summary>
+        private readonly double compactionPercentage;
+
+        /// <summary>Lock to protect access to <see cref="keys"/>.</summary>
+        private readonly object mutex;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MemoryCache{TKey, TValue}"/> class.
+        /// </summary>
+        /// <param name="sizeLimit">Maximum items count that can be stored in the cache.</param>
```

Comment:
if you change a comment or name of these guys, above don't forget to do it here as well

Reply:
Fixed.